src/Mobs/Villager.h

```diff
@@ -3,37 +3,123 @@
 
 #include "PassiveMonster.h"
 #include "Blocks/ChunkInterface.h"
+#include "../UI/WindowOwner.h"
+
+
+
+
+
+struct VillagerTradeOffer
+{
+	VillagerTradeOffer(const cItem & PrimaryTradeDesire, const cItem & TradeRecompense) :
+		PrimaryDesire(PrimaryTradeDesire),
+		Recompense(TradeRecompense),
+		Transactions(0),
+		MaximumTransactions(10)
+	{
+	}
+
+	VillagerTradeOffer(const cItem & PrimaryTradeDesire, const cItem & SecondaryTradeDesire, const cItem & TradeRecompense) :
+		PrimaryDesire(PrimaryTradeDesire),
+		SecondaryDesire(SecondaryTradeDesire),
+		Recompense(TradeRecompense),
+		Transactions(0),
+		MaximumTransactions(10)
+	{
+	}
+
+	cItem PrimaryDesire;
+	cItem SecondaryDesire;
+	cItem Recompense;
+
+	bool IsTradeExhausted() const
+	{
+		return Transactions == MaximumTransactions;
+	}
+
+	unsigned Transactions;
+	unsigned MaximumTransactions;
+};
+
 
 
 
 
 class cVillager :
-	public cPassiveMonster
+	public cPassiveMonster,
+	public cEntityWindowOwner
 {
 	typedef cPassiveMonster super;
 
 public:
 
-	enum eVillagerType
+	enum class VillagerProfession
+	{
+		Farmer = 0,
+		Librarian = 1,
+		Priest = 2,
+		Blacksmith = 3,
+		Butcher = 4,
+		Nitwit = 5
+	};
+
+	enum class VillagerCareer
 	{
-		vtFarmer     = 0,
-		vtLibrarian  = 1,
-		vtPriest     = 2,
-		vtBlacksmith = 3,
-		vtButcher    = 4,
-		vtGeneric    = 5,
-		vtMax
-	} ;
+		Farmer,
+		Fisherman,
+		Shepherd,
+		Fletcher,
+		Librarian,
+		Cartographer,
+		Cleric,
+		Armorer,
+		WeaponSmith,
+		ToolSmith,
+		Butcher,
+		Leatherworker,
+		Nitwit
+	};
+
+	cVillager(VillagerCareer, unsigned);
+	CLASS_PROTODEF(cVillager)
 
-	cVillager(eVillagerType VillagerType);
+	/** Converts internal career to equivalent Vanilla profession.
+	Returns an enumeration with Vanilla-compatible values. */
+	static VillagerProfession VillagerCareerToProfession(VillagerCareer);
 
-	CLASS_PROTODEF(cVillager)
+	/** Returns the career of the villager. */
+	VillagerCareer GetCareer() const { return m_Career; }
+
+	/** Returns a trading offer of the villager subject to criteria.
+	If the player has selected the first trade offer, the first available trade offer matching the input is returned, or the first trade offer if nothing was found.
+	For any other offer selection, the selected offer is returned. */
```

Comment:
This is wrong @tigerw . Example: The villager has a list with three 3 trade offers. When the player right-clicks the villager the current index is at 0. If the player clicks on the arrow (left, right side) the client sends the packet [MC|TrSel](http://wiki.vg/Plugin_channels#MC.7CTrSel) with a number that is the index of the list of the trade offers.

Reply:
ok.

@Seadragon91 I didn't understand what you meant - I believe that that was the behaviour I aimed to implement. The TrSel plugin message is handled elsewhere, and its value is passed into this function as the "selected offer" the comment talks about.

Also, I realised after doing this that you were also working on trading, so now I appear to have duplicated your work, sorry :/

Reply:
I thought you would loop over the trade offers, because of that comment or ist it outdated?: `the first available trade offer matching the input is returned`.

Reply:
For that line in the comment, the corresponding code attempts to replicate what Pokechu was talking about, showing whatever offer in the entire list is available if the inputs match and the player is on the first trade offer.

Which bit do you think is incorrect?